diffrax/_progress_bar.py

```diff
@@ -0,0 +1,186 @@
+import abc
+
+import equinox as eqx
+import jax
+import jax.numpy as jnp
+import numpy as np
+from jax import Array
+from jax.experimental import io_callback
+from tqdm import tqdm
+
+from diffrax._custom_types import FloatScalarLike, IntScalarLike
+
+
+class AbstractProgressBar(eqx.Module):
+    @abc.abstractmethod
+    def new(self) -> IntScalarLike:
+        """Creates a new progress bar and returns its unique identifier"""
+
+    @abc.abstractmethod
+    def update(
+        self,
+        pb_id: IntScalarLike,
+        current_progress: FloatScalarLike,
+        previous_progress: FloatScalarLike,
+        current_step: IntScalarLike,
+    ) -> FloatScalarLike:
+        """Updates the progress bar and returns the last printed progress value"""
+
+    @abc.abstractmethod
+    def close(self, pb_id: IntScalarLike):
+        """Closes the progress bar"""
+
+
+class NoProgressBar(AbstractProgressBar):
+    def new(self) -> IntScalarLike:
+        return jnp.array(0, dtype=jnp.int32)
+
+    def update(
+        self,
+        pb_id: IntScalarLike,
+        current_progress: FloatScalarLike,
+        previous_progress: FloatScalarLike,
+        current_step: IntScalarLike,
+    ) -> FloatScalarLike:
+        return current_progress
+
+    def close(self, pb_id: IntScalarLike):
+        pass
+
+
+class TextProgressBar(AbstractProgressBar):
+    minimum_increase: float
+
+    def __init__(self, minimum_increase: float = 2.0):
+        """
+        Args:
+            minimum_increase: minimum increase of progress bar in percent
+        """
+        self.minimum_increase = minimum_increase
+
+    def new(self) -> IntScalarLike:
+        return jnp.array(0, dtype=jnp.int32)
+
+    def update(
+        self,
+        pb_id: IntScalarLike,
+        current_progress: FloatScalarLike,
+        previous_progress: FloatScalarLike,
+        current_step: IntScalarLike,
+    ) -> FloatScalarLike:
+        """
+        Args:
+            pb_id: progress bar id
+            current_progress: current progress in percent
+            previous_progress: previous progress in percent
+            current_step: current step number
+        """
+
+        def update_aux(progress):
+            # currently
+            # jax.debug.print("{progress:.2f}%", progress=progress)
+            # does not work, so we have to round the progress ourselves
+            jax.debug.print("{}%", np.round(progress, 1), ordered=True)
```

Comment:
Why does this work? `np.round` should be called on NumPy arrays, not JAX arrays?

Reply:
I beleive there is interoperability between numpy and jax.numpy functions, as long as you are not in a jitted piece of code
```python
import jax.numpy as jnp
import numpy as np

a = jnp.linspace(0, 10, 101)
print(jnp.sum(a)) # 505.0
print(np.sum(a)) # 505.0
```

Reply:
Hmm, but isn't all of this code called inside JIT?

Reply:
If I understand correclty it works because [internally](https://github.com/numpy/numpy/blob/d35cd07ea997f033b2d89d349734c61f5de54b0d/numpy/core/fromnumeric.py#L3360) `np.round(a)` calls `a.round()`. To remove all doubt and since is the only use of `np` in the file, I switched to `jnp`